system-addon/lib/TopStoriesFeed.jsm

```diff
@@ -0,0 +1,159 @@
+const TOPICS_UPDATE_TIME = 3 * 60 * 60 * 1000; // 3 hours
+const SECTION_ID = "TopStories";
+
+this.TopStoriesFeed = class TopStoriesFeed {
+  constructor() {
+    this.storiesLastUpdated = 0;
+    this.topicsLastUpdated = 0;
+  }
+
+  init() {
+    try {
+      const prefs = new Prefs();
+      const options = JSON.parse(prefs.get("feeds.topstories.options"));
+      const apiKey = this._getApiKeyFromPref(options.api_key_pref);
+      this.stories_endpoint = this._produceUrlWithApiKey(options.stories_endpoint, apiKey);
+      this.topics_endpoint = this._produceUrlWithApiKey(options.topics_endpoint, apiKey);
+
+      const sectionOptions = {id: SECTION_ID, title: {id: "header_recommended_by", values: {provider: options.provider_name}}};
+      this.store.dispatch(ac.BroadcastToContent({type: at.SECTION_REGISTER, data: sectionOptions}));
+
+      this.fetchStories();
+      this.fetchTopics();
+    } catch (e) {
+      Cu.reportError(`Problem initializing top stories feed: ${e.message}`);
+    }
+  }
+
+  uninit() {
+    this.store.dispatch(ac.BroadcastToContent({type: at.SECTION_DEREGISTER, data: SECTION_ID}));
+  }
+
+  async fetchStories() {
+    if (this.stories_endpoint) {
+      const stories = await fetch(this.stories_endpoint)
+        .then(response => {
+          if (response.ok) {
+            return response.text();
+          }
+          throw new Error(`Stories endpoint returned unexpected status: ${response.status}`);
+        })
+        .then(body => {
+          let items = JSON.parse(body).list;
+          items = items
+            .filter(s => !NewTabUtils.blockedLinks.isBlocked(s.dedupe_url))
+            .map(s => ({
+              "guid": s.id,
+              "recommended": true,
+              "title": s.title,
+              "description": s.excerpt,
+              "bestImage": {"url": this._normalizeUrl(s.image_src)},
+              "url": s.dedupe_url,
+              "lastVisitDate": s.published_timestamp
+            }));
+          return items;
+        })
+        .catch(error => Cu.reportError(`Failed to fetch content: ${error.message}`));
+
+      if (stories) {
+        this.dispatchUpdateEvent(this.storiesLastUpdated,
+          {"type": at.SECTION_ROWS_UPDATE, "data": {"id": SECTION_ID, "rows": stories}});
+        this.storiesLastUpdated = Date.now();
+      }
+    }
+  }
+
+  async fetchTopics() {
+    if (this.topics_endpoint) {
+      const topics = await fetch(this.topics_endpoint)
+        .then(response => {
+          if (response.ok) {
+            return response.text();
+          }
+          throw new Error(`Topics endpoint returned unexpected status: ${response.status}`);
+        })
+        .then(body => JSON.parse(body).topics)
+        .catch(error => Cu.reportError(`Failed to fetch topics: ${error.message}`));
+
+      if (topics) {
+        this.dispatchUpdateEvent(this.topicsLastUpdated,
+          {"type": at.SECTION_ROWS_UPDATE, "data": {"id": SECTION_ID, "topics": topics}});
```

Comment:
This currently won't work as I didn't anticipate anything other than rows being updated with this event but on reflection it's better if it can handle the general case of any data being updated. The fix would be a one-line change in the reducer (line 194).
-- `return Object.assign({}, section, {initialized: true, rows: action.data.rows});`
++ `return Object.assign({}, section, action.data, {initialized: section.rows || action.data.rows});`
We can consider renaming the event to `SECTION_UPDATE`as well but that is less important. Could also put off this fix until we wire all the parts together.